test/common/firefox-cdp-driver.js

```diff
@@ -262,52 +261,52 @@ function setupFrameTracking(client) {
                 break;
             }
         }
+
+        // Firefox does not report an error when the execution context
+        // of a Runtime.evaluate call gets destroyed.  It will never
+        // ever resolve or be rejected.  So let's provide the failure
+        // reply from here.
+        //
+        // However, if the timing is just right, the context gets
+        // destroyed before Runtime.evaluate has started the real
+        // processing, and in that case it will return an error.  Then
+        // we would send the reply here, and would also send the
+        // error. This would drive the protocol out of synch. Also, our driver
+        // might immediately send more commands after seeing the first reply,
+        // and the unwanted second reply might be triggered in the middle of one
+        // of the next commands.  To reliably suppress the second reply we have
+        // the pretty general sequence number checks.
+        //
+        if (info.executionContextId == currentExecId) {
+            currentExecId = null;
+            fail(cur_cmd_seq, { "response": { "message": "Execution context was destroyed." } });
```

Comment:
I'm sure you thought this through, I just find it rather hard to understand. This is an event handler that tells you "an execution context got destroyed", which happens all the time. Raising a test failure here, plus this `currentExecId` is quite a big layering violation and hard to keep track of in one's head. See below for a proposal.

Reply:
It's a complicated workaround for a bug in Firefox's CDP.  But as long as it stays in the driver here, I wouldn't complain.